homeassistant/components/refoss/__init__.py

```diff
@@ -0,0 +1,130 @@
+            entry.title,
+        )
+        return False
+
+    hass.data.setdefault(DOMAIN, {})
+
+    device_ids: set[str] = set()
+    socketserver: SocketServerProtocol = await get_refoss_socket_server(hass)
+    device_manager = RefossDeviceManager(socket_server=socketserver)
+    await device_manager.async_start_broadcast_msg()
+    listener = DeviceListener(hass, device_manager, device_ids)
+    device_manager.add_device_listener(listener)
+
+    hass.data[DOMAIN][entry.entry_id] = HomeAssistantRefossData(
+        device_listener=listener,
+        device_manager=device_manager,
+    )
+
+    await cleanup_device_registry(hass, device_manager)
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    return True
+
+
+async def cleanup_device_registry(
+    hass: HomeAssistant, device_manager: RefossDeviceManager
+) -> None:
+    """Remove deleted device registry entry if there are no remaining entities."""
+    device_registry = dr.async_get(hass)
+    for dev_id, device_entry in list(device_registry.devices.items()):
+        for item in device_entry.identifiers:
+            if item[0] == DOMAIN and item[1] not in device_manager.base_device_map:
+                device_registry.async_remove_device(dev_id)
+                break
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unloading the refoss platforms."""
+    unload = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+    if unload:
+        hass_data: HomeAssistantRefossData = hass.data[DOMAIN][entry.entry_id]
+        hass_data.device_manager.remove_device_listener(hass_data.device_listener)
+        hass_data.device_manager.base_device_map.clear()
+        for task in hass_data.device_manager.tasks:
+            task.cancel()
+
+        hass.data[DOMAIN].pop(entry.entry_id)
+        if not hass.data[DOMAIN]:
+            hass.data.pop(DOMAIN)
+    return unload
+
+
+class DeviceListener(RefossDeviceListener):
+    """Device Update Listener."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        device_manager: RefossDeviceManager,
+        device_ids: set[str],
+    ) -> None:
+        """Init DeviceListener."""
+        self.hass = hass
+        self.device_manager = device_manager
+        self.device_ids = device_ids
+
+    def update_device(self, device: BaseDevice) -> None:
+        """Update device status."""
+        if device.uuid in self.device_ids:
+            dispatcher_send(
+                self.hass, f"{REFOSS_HA_SIGNAL_UPDATE_ENTITY}_{device.uuid}"
+            )
+
+    def add_device(self, device: BaseDevice) -> None:
+        """Add device."""
+        self.async_remove_device(device.uuid)
+
+        self.device_ids.add(device.uuid)
+        dispatcher_send(self.hass, REFOSS_DISCOVERY_NEW, [device.uuid])
+        dispatcher_send(self.hass, f"{REFOSS_HA_SIGNAL_UPDATE_ENTITY}_{device.uuid}")
+        LOGGER.debug("Add device: %s", device.device_type)
+
```

Comment:
It looks like this is use the sync api. It would be better if the library was fully async since this is a new integration

Reply:
Changed to async